File: apps/dashboard/src/api/pipelines.ts
import { api } from "./client";
import type {
  Pipeline,
  PipelineMode,
  PipelineStage,
  PipelineStatus,
  StageStatus,
} from "@/types/pipeline";

interface ResponseEnvelope<T> {
  status: "success";
  data: T;
  meta: { request_id: string; timestamp: string };
}

interface PaginatedEnvelope<T> {
  status: "success";
  data: T[];
  meta: { request_id: string; timestamp: string };
  pagination: {
    page: number;
    per_page: number;
    total: number;
    total_pages: number;
  } | null;
}

interface BackendPipelineResponse {
  id: string;
  project_id: string;
  status: PipelineStatus;
  current_phase: string;
  started_at: string | null;
  completed_at: string | null;
  total_tokens_used: number;
  total_cost_usd: number;
  graph_definition?: {
    mode?: string;
    config?: Record<string, unknown> | null;
  } | null;
  error_message?: string | null;
  config?: Record<string, unknown> | null;
  mode?: string | null;
}

interface BackendPipelinePhaseResponse {
  id: string;
  name: string;
  phase_type: string;
  status: string;
  order: number;
  requires_approval: boolean;
  approved_by: string | null;
  started_at: string | null;
  completed_at: string | null;
  error_message?: string | null;
}

interface PhaseApprovalPayload {
  approved: boolean;
  comment?: string;
}

interface PipelineActionResponse {
  id: string;
  status: PipelineStatus;
  timestamp: string;
}

function normalizeStageStatus(status: string): StageStatus {
  switch (status) {
    case "running":
      return "running";
    case "completed":
      return "completed";
    case "failed":
      return "failed";
    case "skipped":
      return "skipped";
    case "waiting_approval":
      return "waiting";
    case "pending":
    default:
      return "idle";
  }
}

function normalizeMode(mode: string | null | undefined): PipelineMode {
  const normalized = (mode ?? "full").replace(/-/g, "_");
  if (normalized === "quick" || normalized === "review_only") {
    return normalized;
  }
  return "full";
}

function normalizeStages(phases: BackendPipelinePhaseResponse[]): PipelineStage[] {
  return phases.map((phase) => ({
    id: phase.id,
    name: phase.name,
    stage_number: phase.order,
    status: normalizeStageStatus(phase.status),
    started_at: phase.started_at,
    completed_at: phase.completed_at,
    agents: [],
    phase_type: phase.phase_type,
    requires_approval: phase.requires_approval,
    approved_by: phase.approved_by,
    error_message: phase.error_message ?? null,
  }));
}

export function normalizePipeline(
  pipeline: BackendPipelineResponse,
  phases: BackendPipelinePhaseResponse[] = [],
): Pipeline {
  return {
    id: pipeline.id,
    project_id: pipeline.project_id,
    mode: normalizeMode(pipeline.mode ?? pipeline.graph_definition?.mode),
    status: pipeline.status,
    current_phase: pipeline.current_phase,
    total_stages: phases.length,
    stages: normalizeStages(phases),
    config: pipeline.config ?? pipeline.graph_definition?.config ?? null,
    started_at: pipeline.started_at,
    completed_at: pipeline.completed_at,
    total_tokens_used: pipeline.total_tokens_used,
    total_cost_usd: pipeline.total_cost_usd,
    error_message: pipeline.error_message ?? null,
  };
}

export const pipelineApi = {
  async list(projectId: string) {
    const response = await api.get<PaginatedEnvelope<BackendPipelineResponse>>(
      `/api/v1/projects/${projectId}/pipelines`,
    );
    return {
      ...response,
      data: response.data.map((pipeline) => normalizePipeline(pipeline)),
    };
  },

  async get(pipelineId: string) {
    const [pipelineResponse, phasesResponse] = await Promise.all([
      api.get<ResponseEnvelope<BackendPipelineResponse>>(
        `/api/v1/pipelines/${pipelineId}`,
      ),
      api.get<ResponseEnvelope<BackendPipelinePhaseResponse[]>>(
        `/api/v1/pipelines/${pipelineId}/phases`,
      ),
    ]);

    return {
      ...pipelineResponse,
      data: normalizePipeline(pipelineResponse.data, phasesResponse.data),
    };
  },

  async create(projectId: string, mode: PipelineMode) {
    const response = await api.post<ResponseEnvelope<BackendPipelineResponse>>(
      `/api/v1/projects/${projectId}/pipelines`,
      { mode },
    );
    return {
      ...response,
      data: normalizePipeline(response.data),
    };
  },

  start(pipelineId: string) {
    return api.post<ResponseEnvelope<PipelineActionResponse>>(
      `/api/v1/pipelines/${pipelineId}/start`,
    );
  },

  pause(pipelineId: string) {
    return api.post<ResponseEnvelope<PipelineActionResponse>>(
      `/api/v1/pipelines/${pipelineId}/pause`,
    );
  },

  resume(pipelineId: string) {
    return api.post<ResponseEnvelope<PipelineActionResponse>>(
      `/api/v1/pipelines/${pipelineId}/resume`,
    );
  },

  cancel(pipelineId: string) {
    return api.post<ResponseEnvelope<PipelineActionResponse>>(
      `/api/v1/pipelines/${pipelineId}/cancel`,
    );
  },

  approvePhase(
    pipelineId: string,
    phaseId: string,
    payload: PhaseApprovalPayload,
  ) {
    return api.post<ResponseEnvelope<BackendPipelinePhaseResponse>>(
      `/api/v1/pipelines/${pipelineId}/phases/${phaseId}/approve`,
      payload,
    );
  },
};

File: apps/server/src/codebot/api/routes/pipelines.py
"""Pipeline CRUD and lifecycle endpoints for the CodeBot API."""

import math
from datetime import UTC, datetime
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from codebot.api.deps import get_current_user, get_db, require_role
from codebot.api.envelope import PaginatedEnvelope, PaginationMeta, ResponseEnvelope
from codebot.api.schemas.pipelines import (
    PhaseApprovalRequest,
    PipelineActionResponse,
    PipelineCreate,
    PipelineDetailResponse,
    PipelinePhaseResponse,
    PipelineResponse,
)
from codebot.db.models.project import PipelineStatus
from codebot.db.models.user import User
from codebot.services.pipeline_service import PipelineService

router = APIRouter(prefix="/pipelines", tags=["pipelines"])
project_pipelines_router = APIRouter(prefix="/projects", tags=["pipelines"])


def _get_pipeline_service(
    db: Annotated[AsyncSession, Depends(get_db)],
) -> PipelineService:
    """Dependency that provides a PipelineService instance."""
    return PipelineService(db)


# ---------------------------------------------------------------------------
# Project-scoped pipeline endpoints (nested under /projects/{project_id})
# ---------------------------------------------------------------------------


@project_pipelines_router.post(
    "/{project_id}/pipelines",
    response_model=ResponseEnvelope[PipelineResponse],
    status_code=status.HTTP_201_CREATED,
)
async def create_pipeline(
    project_id: UUID,
    body: PipelineCreate,
    current_user: Annotated[User, Depends(require_role("user", "admin"))],
    service: Annotated[PipelineService, Depends(_get_pipeline_service)],
) -> ResponseEnvelope[PipelineResponse]:
    """Create a new pipeline for a project.

    Args:
        project_id: The project UUID.
        body: Pipeline creation data (mode, phases, config).
        current_user: Authenticated user (user or admin role).
        service: Pipeline service dependency.

    Returns:
        Created pipeline wrapped in response envelope.
    """
    pipeline = await service.create(project_id, body)
    return ResponseEnvelope(data=PipelineResponse.model_validate(pipeline))


@project_pipelines_router.get(
    "/{project_id}/pipelines",
    response_model=PaginatedEnvelope[PipelineResponse],
)
async def list_pipelines(
    project_id: UUID,
    current_user: Annotated[User, Depends(get_current_user)],
    service: Annotated[PipelineService, Depends(_get_pipeline_service)],
    page: int = Query(default=1, ge=1),
    per_page: int = Query(default=20, ge=1, le=100),
    status_filter: str | None = Query(default=None, alias="status"),
) -> PaginatedEnvelope[PipelineResponse]:
    """List pipelines for a project.

    Args:
        project_id: The project UUID.
        current_user: Authenticated user.
        service: Pipeline service dependency.
        page: Page number (1-based).
        per_page: Items per page.
        status_filter: Optional status filter.

    Returns:
        Paginated list of pipelines.
    """
    pipelines, total = await service.list_for_project(
        project_id, page=page, per_page=per_page, status=status_filter
    )
    total_pages = math.ceil(total / per_page) if per_page > 0 else 0
    return PaginatedEnvelope(
        data=[PipelineResponse.model_validate(p) for p in pipelines],
        pagination=PaginationMeta(
            page=page, per_page=per_page, total=total, total_pages=total_pages
        ),
    )


# ---------------------------------------------------------------------------
# Pipeline-level endpoints (direct /pipelines/{pipeline_id})
# ---------------------------------------------------------------------------


@router.get(
    "/{pipeline_id}",
    response_model=ResponseEnvelope[PipelineDetailResponse],
)
async def get_pipeline(
    pipeline_id: UUID,
    current_user: Annotated[User, Depends(get_current_user)],
    service: Annotated[PipelineService, Depends(_get_pipeline_service)],
) -> ResponseEnvelope[PipelineDetailResponse]:
    """Get a pipeline by ID with full detail.

    Args:
        pipeline_id: The pipeline UUID.
        current_user: Authenticated user.
        service: Pipeline service dependency.

    Returns:
        Pipeline detail wrapped in response envelope.
    """
    pipeline = await service.get_with_phases(pipeline_id)
    if pipeline is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Pipeline not found"
        )
    return ResponseEnvelope(data=PipelineDetailResponse.model_validate(pipeline))


@router.post(
    "/{pipeline_id}/start",
    response_model=ResponseEnvelope[PipelineActionResponse],
)
async def start_pipeline(
    pipeline_id: UUID,
    current_user: Annotated[User, Depends(require_role("user", "admin"))],
    service: Annotated[PipelineService, Depends(_get_pipeline_service)],
) -> ResponseEnvelope[PipelineActionResponse]:
    """Start a pipeline (transition from PENDING to RUNNING).

    Args:
        pipeline_id: The pipeline UUID.
        current_user: Authenticated user (user or admin role).
        service: Pipeline service dependency.

    Returns:
        Pipeline action response.
    """
    pipeline = await service.get(pipeline_id)
    if pipeline is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Pipeline not found"
        )
    updated = await service.transition(pipeline, PipelineStatus.RUNNING)
    return ResponseEnvelope(
        data=PipelineActionResponse(
            id=updated.id,
            status=updated.status.value.lower(),
            timestamp=datetime.now(UTC),
        )
    )


@router.post(
    "/{pipeline_id}/pause",
    response_model=ResponseEnvelope[PipelineActionResponse],
)
async def pause_pipeline(
    pipeline_id: UUID,
    current_user: Annotated[User, Depends(require_role("user", "admin"))],
    service: Annotated[PipelineService, Depends(_get_pipeline_service)],
) -> ResponseEnvelope[PipelineActionResponse]:
    """Pause a running pipeline.

    Args:
        pipeline_id: The pipeline UUID.
        current_user: Authenticated user.
        service: Pipeline service dependency.

    Returns:
        Pipeline action response.
    """
    pipeline = await service.get(pipeline_id)
    if pipeline is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Pipeline not found"
        )
    updated = await service.transition(pipeline, PipelineStatus.PAUSED)
    return ResponseEnvelope(
        data=PipelineActionResponse(
            id=updated.id,
            status=updated.status.value.lower(),
            timestamp=datetime.now(UTC),
        )
    )


@router.post(
    "/{pipeline_id}/resume",
    response_model=ResponseEnvelope[PipelineActionResponse],
)
async def resume_pipeline(
    pipeline_id: UUID,
    current_user: Annotated[User, Depends(require_role("user", "admin"))],
    service: Annotated[PipelineService, Depends(_get_pipeline_service)],
) -> ResponseEnvelope[PipelineActionResponse]:
    """Resume a paused pipeline.

    Args:
        pipeline_id: The pipeline UUID.
        current_user: Authenticated user.
        service: Pipeline service dependency.

    Returns:
        Pipeline action response.
    """
    pipeline = await service.get(pipeline_id)
    if pipeline is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Pipeline not found"
        )
    updated = await service.transition(pipeline, PipelineStatus.RUNNING)
    return ResponseEnvelope(
        data=PipelineActionResponse(
            id=updated.id,
            status=updated.status.value.lower(),
            timestamp=datetime.now(UTC),
        )
    )


@router.post(
    "/{pipeline_id}/cancel",
    response_model=ResponseEnvelope[PipelineActionResponse],
)
async def cancel_pipeline(
    pipeline_id: UUID,
    current_user: Annotated[User, Depends(require_role("user", "admin"))],
    service: Annotated[PipelineService, Depends(_get_pipeline_service)],
) -> ResponseEnvelope[PipelineActionResponse]:
    """Cancel a pipeline.

    Args:
        pipeline_id: The pipeline UUID.
        current_user: Authenticated user.
        service: Pipeline service dependency.

    Returns:
        Pipeline action response.
    """
    pipeline = await service.get(pipeline_id)
    if pipeline is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Pipeline not found"
        )
    updated = await service.transition(pipeline, PipelineStatus.CANCELLED)
    return ResponseEnvelope(
        data=PipelineActionResponse(
            id=updated.id,
            status=updated.status.value.lower(),
            timestamp=datetime.now(UTC),
        )
    )


@router.get(
    "/{pipeline_id}/phases",
    response_model=ResponseEnvelope[list[PipelinePhaseResponse]],
)
async def get_pipeline_phases(
    pipeline_id: UUID,
    current_user: Annotated[User, Depends(get_current_user)],
    service: Annotated[PipelineService, Depends(_get_pipeline_service)],
) -> ResponseEnvelope[list[PipelinePhaseResponse]]:
    """Get all phases of a pipeline.

    Args:
        pipeline_id: The pipeline UUID.
        current_user: Authenticated user.
        service: Pipeline service dependency.

    Returns:
        Ordered list of pipeline phases.
    """
    phases = await service.get_phases(pipeline_id)
    return ResponseEnvelope(
        data=[PipelinePhaseResponse.model_validate(p) for p in phases]
    )


@router.post(
    "/{pipeline_id}/phases/{phase_id}/approve",
    response_model=ResponseEnvelope[PipelinePhaseResponse],
)
async def approve_phase(
    pipeline_id: UUID,
    phase_id: UUID,
    body: PhaseApprovalRequest,
    current_user: Annotated[User, Depends(require_role("user", "admin"))],
    service: Annotated[PipelineService, Depends(_get_pipeline_service)],
) -> ResponseEnvelope[PipelinePhaseResponse]:
    """Approve or reject a pipeline phase.

    Args:
        pipeline_id: The pipeline UUID.
        phase_id: The phase UUID.
        body: Approval request data.
        current_user: Authenticated user (user or admin role).
        service: Pipeline service dependency.

    Returns:
        Updated phase wrapped in response envelope.
    """
    phases = await service.get_phases(pipeline_id)
    phase = next((p for p in phases if p.id == phase_id), None)
    if phase is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Phase not found"
        )
    updated = await service.approve_phase(
        phase,
        approved_by=current_user.email if hasattr(current_user, "email") else str(current_user.id),
        comment=body.comment,
    )
    return ResponseEnvelope(data=PipelinePhaseResponse.model_validate(updated))
